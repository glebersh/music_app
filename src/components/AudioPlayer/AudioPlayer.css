.play-icon {
  font-size: 3em;
  color: white;
}

.next-icon {
  font-size: 2em;
  color: white;
}

audio::-webkit-media-controls-play-button {
  display: none;
}