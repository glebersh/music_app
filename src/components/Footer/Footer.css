footer {
  width: 100%;
  z-index: 2000;
  height: 100px;
  position: fixed;
  margin-top: auto;
  bottom: 0;
  left: 0;
  right: 0;
  background: rgb(20, 20, 20, 0.7);
}