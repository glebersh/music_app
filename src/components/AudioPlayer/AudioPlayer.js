import { Flex, Box, Image, Text } from '@chakra-ui/react';
import { Link } from 'react-router-dom';
import React, { useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { setCurrentSong, setIsPlaying, setNextSong, setPreviousSong, setSongsCollectionType } from '../../store/slices/playerSlice';
import './AudioPlayer.css';
import ErrorBoundary from '../ErrorBoundary/ErrorBoundary';

const AudioPlayer = () => {
  const dispatch = useDispatch();

  const isPlaying = useSelector(state => state.playerReducer.isPlaying);
  const currentSong = useSelector(state => state.playerReducer.currentSong);
  const previousSong = useSelector(state => state.playerReducer.previousSong);
  const nextSong = useSelector(state => state.playerReducer.nextSong);
  const isEmpty = useSelector(state => state.searchReducer.isEmpty);
  const isLoading = useSelector(state => state.searchReducer.isLoading);
  const playlist = useSelector(state => state.playerReducer.songsCollection);
  const collectionType = useSelector(state => state.playerReducer.songsCollectionType);

  const audioPlayer = document.querySelector('#audio');
  // const link = isEmpty && currentSong ? '' : `artists/${currentSong.artists[0].id}`;
  const link = '';

  useEffect(() => {
    if (document.readyState === 'complete') {
      if (isPlaying) {
        audioPlayer.play();
      }
      else if (!isPlaying) {
        audioPlayer.pause();
      }
    }
  }, [isPlaying, currentSong.preview_url]);


  const onPlay = () => {
    dispatch(setIsPlaying(!isPlaying));
  };

  const playPreviousSong = () => {
    if (previousSong.preview_url !== '' && previousSong.preview_url !== currentSong.preview_url) {
      dispatch(setCurrentSong(previousSong));
    }
    if (collectionType === 'albums') {
      if (currentSong.track_number > 0) {
        const previousIndex = currentSong.track_number - 2;
        dispatch(setCurrentSong(playlist.tracks.items[previousIndex]));
      }
    }
    else if (collectionType === 'songs') {
      dispatch(setSongsCollectionType('albums'));
      dispatch(setCurrentSong(previousSong));
    }
  };

  const playNextSong = () => {
    if (collectionType === 'albums') {
      if (currentSong.track_number <= playlist.total_tracks) {
        dispatch(setPreviousSong(currentSong));
        dispatch(setCurrentSong(playlist.tracks.items[currentSong.track_number]));
      }
    };
  }


  const coverImage = !isEmpty && !isLoading && playlist ? collectionType === 'songs' ?
    currentSong.album.images[0].url :
    collectionType === 'albums' ?
      playlist.images[0].url : currentSong.track.album.images[0].url : null;

  const trackName = !isEmpty && !isLoading && playlist ? collectionType === 'songs' ? currentSong.name : currentSong.name : null;

  const previewLink = !isEmpty && !isLoading && playlist ? collectionType === 'songs' ? currentSong.preview_url : currentSong.preview_url : null;

  const artistName = !isEmpty && !isLoading && playlist ? collectionType === 'songs' ? currentSong.album.artists[0].name :
    collectionType === 'albums' ? playlist.artists[0].name : currentSong.track.artists[0].name : null;

  const artistID = !isEmpty && !isLoading && playlist ? collectionType === 'songs' ?
    currentSong.album.artists[0].name :
    collectionType === 'albums' ?
      playlist.artists[0].id : currentSong.track.artists[0].id : null;

  const albumName = !isEmpty && !isLoading && playlist ? collectionType === 'songs' ? currentSong.album.name : playlist.name : null;


  return (
    <ErrorBoundary>
      <Flex justify='space-between' gap='3em' align='center' h='100%'>
        <Flex justify='space-evenly' gap='1.5em' align='center' ml='5em'>
          <Image src={coverImage} fallbackSrc='https://via.placeholder.com/64' maxW='64px' maxH='64px' />
          <Flex direction='column' ml='1em'>
            <Box>
              <Text w='100%' fontWeight='700' display='inline' mr='1em' color='white'>{trackName}</Text>
              {/* <span>{currentSong.explicit ? <i className="bi bi-explicit"></i> : null}</span> */}
            </Box>
            <Text minW='100%' color='white'>{albumName}</Text>
            <Link to={link}><Text color='white'
              _hover={{
                cursor: 'pointer',
                color: 'primary'
              }}>{artistName}</Text></Link>
          </Flex>
        </Flex>
        <Flex justify='flex-start' w='64%'>
          <Flex justify='center' gap='3em' align='center' h='100%'>
            {currentSong.track_number === 1 && collectionType === 'albums' ? null : <i className="bi bi-skip-start next-icon"
              onClick={(e) => playPreviousSong(e)}></i>}

            <audio src={previewLink} id='audio' controls></audio>

            {isPlaying ? <i className="bi bi-pause play-icon" onClick={() => onPlay()}></i> : <i className="bi bi-play-fill play-icon" onClick={() => onPlay()}></i>}

            {currentSong.track_number === playlist.total_tracks & collectionType === 'albums' ? null : <i className="bi bi-skip-end next-icon" id='next-button'
              onClick={(e) => playNextSong(e)}></i>}
          </Flex>
        </Flex>
      </Flex>
    </ErrorBoundary>
  )
};
export default AudioPlayer;