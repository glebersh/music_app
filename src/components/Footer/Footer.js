import React from 'react';
import AudioPlayer from '../AudioPlayer/AudioPlayer';
import './Footer.css';


const Footer = () => {

  return (
    <footer>
      <AudioPlayer />
    </footer>
  )
};

export default Footer;
