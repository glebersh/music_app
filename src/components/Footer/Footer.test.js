import { render } from "@testing-library/react"
import Footer from "./Footer"
import '@testing-library/jest-dom';
import * as AudioPlayer from '../AudioPlayer/AudioPlayer';
import * as reduxHooks from 'react-redux';
import React from "react";

jest.mock('react-redux');
const mockedUseSelector = jest.spyOn(reduxHooks, 'useSelector');

jest.mock('../AudioPlayer/AudioPlayer', () => () => (
  <div>Audio Player with no functionality</div>
));


describe('Footer list tests', () => {
  it('Should render and match snapshot', () => {
    const footer = render(<Footer />);
    expect(footer).toMatchSnapshot();
  });
});