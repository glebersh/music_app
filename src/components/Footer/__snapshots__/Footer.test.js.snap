// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer list tests Should render and match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <footer>
        <div
          class="css-7pf6at"
        >
          <div>
            Audio Player with no functionality
          </div>
          <button
            class="chakra-button css-vo7oeh"
            type="button"
          >
            Show tracklist
          </button>
        </div>
      </footer>
    </div>
  </body>,
  "container": <div>
    <footer>
      <div
        class="css-7pf6at"
      >
        <div>
          Audio Player with no functionality
        </div>
        <button
          class="chakra-button css-vo7oeh"
          type="button"
        >
          Show tracklist
        </button>
      </div>
    </footer>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
